### PR TITLE
8462 compare cut

### PR DESCRIPTION
**Pull request by gplsteph**
First shoot at being able to compare Cuts:
- Introduced `SGCutDiff`, `SGCutDiffGroup`, `SGTrackDiff` classes.
- Added `omitted_statuses` settings which needs to be revisited (needs to be SG Status Entities?).
- Reproduced code from Import Cut with them.
- Added some helpers, refactored a couple of things to be able to re-use code.
- Added Cut diff tests.

Still to do: add tests for all various cut changes.

### CODE REVIEW

**Comment by gplsteph** [comment]
Tests are failing because our `.flake8` is not taken into consideration :(

File: sg_otio/track_diff.py [hunk 1]
```diff
@@ -0,0 +1,395 @@
...
+            for sg_shot in sg_more_shots:
+                shot_name = sg_shot["code"].lower()
+                self._diffs_by_shots[shot_name].sg_shot = sg_shot
+
+        # Duplicate the list of shots, allowing us to know easily which ones are
+        # not part of the new track by removing entries when we use them.
+        # We only need a shallow copy here
+        leftover_shots = [x for x in sg_shots_dict.values()]
+        seen_names = []
+        duplicate_names = {}
```
Comment (ksallee) [inline sg_otio/track_diff.py:134]: maybe use a `defaultdict(int)`?

File: sg_otio/track_diff.py [hunk 2]
```diff
@@ -0,0 +1,395 @@
...
+        - Is it linked to the right Version?
+        - Is the index the same?
+        - Is the tc in the same?
+        - Is the tc out the same?
+
+        :param for_clip: A Clip instance.
+        :param prev_clip_list: A list of Clip instances to consider.
+        :param sg_shot: A SG Shot dictionary
+        :param sg_version: A SG Version dictionary
+        :returns: A Clip, or ``None``
```
Comment (ksallee) [inline sg_otio/track_diff.py:265]:
```suggestion
        :param sg_shot: A SG Shot dictionary.
        :param sg_version: A SG Version dictionary.
        :returns: A Clip, or ``None``.
```

File: sg_otio/track_diff.py [hunk 3]
```diff
@@ -0,0 +1,395 @@
...
+        if diff == 0:
+            score += 1
+
+        return score
+
+    def __len__(self):
+        """
+        Return the total number of :class:`SGCutDiff` entries in this SGTrackDiff.
+
+        :returns: An integer
```
Comment (ksallee) [inline sg_otio/track_diff.py:359]:
```suggestion
        :returns: An integer.
```

File: sg_otio/track_diff.py [hunk 4]
```diff
@@ -0,0 +1,395 @@
...
+        """
+        return sum([len(self._diffs_by_shots[k]) for k in self._diffs_by_shots], 0)
+
+    def __iter__(self):
+        """
+        Iterate over Shots for this SGTrackDiff
+
+        :yields: Shot names, as strings
```
Comment (ksallee) [inline sg_otio/track_diff.py:367]:
```suggestion
        Iterate over Shots for this SGTrackDiff.

        :yields: Shot names, as strings.
```

File: sg_otio/track_diff.py [hunk 5]
```diff
@@ -0,0 +1,395 @@
...
+        if key:
+            return self._diffs_by_shots.get(key.lower())
+        return self._diffs_by_shots.get(key)
+
+    def iteritems(self):
+        """
+        Iterate over Shot names for this SGTrackDiff, yielding (name, :class:`SGCutDiffGroup`)
+        tuple
+
+        :yields: (name, :class:`SGCutDiffGroup`) tuples
```
Comment (ksallee) [inline sg_otio/track_diff.py:388]:
```suggestion
        Iterate over Shot names for this SGTrackDiff, yielding (name, :class:`SGCutDiffGroup`)
        tuples.

        :yields: (name, :class:`SGCutDiffGroup`) tuples.
```

**Review by ksallee — APPROVED** [review]
LGTM!

File: sg_otio/cut_diff.py [hunk 6]
```diff
@@ -0,0 +1,319 @@
...
+        # No current clip
+        if self._as_omitted:
+            if not self.repeated:
+                self._diff_type = _DIFF_TYPES.OMITTED
+            else:
+                self._diff_type = _DIFF_TYPES.OMITTED_IN_CUT
+            return
+
+        # We have both a Shot and a current clip.
+        if self.sg_shot_status and self.sg_shot_status.get("code"):
```
[the remark this answers is not in the document]
Comment (gplsteph) [reply sg_otio/cut_diff.py:224]: No it's a dictionary. I have an extra ticket to clarify what we should do for this.

File: sg_otio/track_diff.py (hunk 1 above)
Comment (gplsteph) [reply sg_otio/track_diff.py:134]: will keep it like this if you don't mind.

**Comment by gplsteph** [comment]
@ksallee please have another look. There are still a couple of things missing, but we're getting close!

File: bin/sg-otio.py [hunk 7]
```diff
@@ -0,0 +1,314 @@
...
+    read_parser = subparser.add_parser(
+        "read",
+        help="Read a Cut from SG and write it to a file, or to the std output in OTIO format."
+    )
+    # Write sub parser
+    write_parser = subparser.add_parser(
+        "write",
+        help="Write a Timeline or a video track to SG as a Cut."
+    )
+    # Write sub parser
```
Comment (ksallee) [inline bin/sg-otio.py:47]:
```suggestion
    # Compare sub parser
```

File: bin/sg-otio.py [hunk 8]
```diff
@@ -0,0 +1,314 @@
...
+    compare_parser = subparser.add_parser(
+        "compare",
+        help="Compare a Timeline or a video track to a SG Cut."
+    )
+    # Common arguments for read and write.
```
Comment (ksallee) [inline bin/sg-otio.py:52]:
```suggestion
    # Common arguments for read, write and compare.
```

File: bin/sg-otio.py [hunk 9]
```diff
@@ -0,0 +1,314 @@
...
+        sg_site_url=args.sg_site_url,
+        cut_id=args.cut_id,
+        session_token=_get_session_token(args)
+    )
+    timeline = otio.adapters.read_from_file(
+        url,
+        adapter_name="ShotGrid",
+    )
+    if not args.file:
+        logger.info(otio.adapters.write_to_string(timeline))
```
Comment (ksallee) [inline bin/sg-otio.py:150]: shouldn't we add here also `adapter_name=args.adapter`?

File: bin/sg-otio.py [hunk 10]
```diff
@@ -0,0 +1,314 @@
...
+    :param args: The command line arguments.
+    """
+    session_token = _get_session_token(args)
+    url = get_write_url(
+        sg_site_url=args.sg_site_url,
+        entity_type=args.entity_type,
+        entity_id=args.entity_id,
+        session_token=session_token
+    )
+    timeline = otio.adapters.read_from_file(
```
Comment (ksallee) [inline bin/sg-otio.py:171]: I think there's something missing here for EDLs. If they're not 24fps, normally you need to provide the frame rate. I think we need to add an argument and add it here if provided. Maybe we should create a ticket for this?

File: tests/test_cut_diff.py [hunk 11]
```diff
@@ -0,0 +1,621 @@
...
+                self.assertEqual(
+                    cut_diff.current_clip.index, cut_diff.old_clip.index
+                )
+                self.assertEqual(
+                    cut_diff.cut_in, cut_diff.old_cut_in
+                )
+                self.assertEqual(
+                    cut_diff.cut_out, cut_diff.old_cut_out
+                )
+                logger.info(cut_diff.reasons)
```
Comment (ksallee) [inline tests/test_cut_diff.py:189]: do we need this?

**Review by ksallee — APPROVED** [review]
LGTM

File: bin/sg-otio.py (hunk 10 above)
Comment (gplsteph) [reply bin/sg-otio.py:171]: Ticket here: https://gpltech.shotgunstudio.com/detail/Ticket/9322

File: tests/test_cut_diff.py (hunk 11 above)
Comment (gplsteph) [reply tests/test_cut_diff.py:189]: No :)

**Review by ksallee — DISMISSED** [review]
LGTM, but I have some questions, let's talk!

**Review by ksallee — APPROVED** [review]
LGTM

**Pull request merged**